galsim/catalog.py

```diff
@@ -312,3 +317,334 @@ def itervalues(self):
+    right pixel values to mimic the actual COSMOS image.
+
+    @param cat        The output from makeCOSMOSCatalog
+    @param index      The index of the object of interest in this catalog.
+    @param chromatic  Make this a chromatic object, or not?  [default: False]
+                      It is important to bear in mind that we do not actually have
+                      spatially-resolved color information for these galaxies, so this keyword can
+                      only be True if we are using parametric galaxies.  Even then, we simply do the
+                      most arbitrary thing possible, which is to assign bulges an elliptical SED,
+                      disks a disk-like SED, and Sersic galaxies with intermediate values of n some
+                      intermediate SED.  We assume that the photometric redshift is the correct
+                      redshift for these galaxies (which is a good assumption for COSMOS 30-band
+                      photo-z for these bright galaxies).  For the given SED and redshift, we then
+                      normalize to give the right (observed) flux in F814W.  Note that for a mock
+                      "deep" sample, the redshift distributions of the galaxies would be modified,
+                      which is not included here.
+    @param pad_size   For realistic galaxies, the size of region requiring noise padding, in arcsec.
+
+    @returns Either a GSObject or a chromatic object representing the galaxy of interest.
+    """
+    # Check whether this is a catalog entry for a real object or for a parametric one.
+    if isinstance(cat, galsim.RealGalaxyCatalog):
+        if pad_size is None:
+            pad_size=0.25 # random and not completely ridiculous guess in arcsec
+        if chromatic:
+            raise RuntimeError("Cannot yet make real chromatic galaxies!")
+        return _makeReal(cat, index, pad_size=pad_size)
+    else:
+        return _makeParam(cat, index, chromatic=chromatic)
+
+def _makeReal(cat, index, pad_size):
+    noise_pad_size = int(np.ceil(pad_size * np.sqrt(2.)))
+    gal = galsim.RealGalaxy(cat, index=index, noise_pad_size=noise_pad_size)
+
+    # Rescale the galaxy size.
+    if hasattr(cat, 'size_factor'):
+        gal.applyDilation(cat.size_factor)
+    # Rescale the galaxy flux.
+    if hasattr(cat, 'flux_factor'):
+        gal *= cat.flux_factor
+    return gal
+
+# This will act essentially as a static variable for the _makeParam function
+_COSMOS_bandpass = None
+
+def _makeParam(cat, index, chromatic=False):
+    if _COSMOS_bandpass is None:
+        # Defer making the Bandpass until we actually use it.
+        # It's not a huge calculation, but the thin() call especially isn't trivial.
+        _COSMOS_bandpass = galsim.Bandpass(
+                os.path.join(galsim.meta_data.share_dir, 'wfc_F814W.dat.gz'),
+                wave_type='ang').thin().withZeropoint(25.94)
```

Comment:
I moved the loading of the bandpass to here where it is actually used.  cf. 83df8e2

The thin() function isn't all that cheap, so doing this every time someone imports galsim is not a good idea.  This is design pattern is kind of the python equivalent of static variables in C++.